### PR TITLE
Support MessagePack for persistent mode

### PR DESCRIPTION
**Pull request by Y-Sindo**
Add method `ServiceManagerBuilder.WithHubProtocols`, `ServiceManagerBuilder.AddHubProtocol`

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs [hunk 1]
```diff
@@ -79,14 +80,40 @@ public ServiceManagerBuilder WithNewtonsoftJson(Action<NewtonsoftServiceHubProto
...
+        /// <param name="hubProtocols">Only the protocols named "json" or "messagepack" are allowed.</param>
+        /// <returns>The <see cref="ServiceHubContextBuilder"/> instance itself.</returns>
+        public ServiceManagerBuilder WithHubProtocol(IEnumerable<IHubProtocol> hubProtocols)
+        {
+            if (hubProtocols == null)
+            {
+                throw new ArgumentNullException(nameof(hubProtocols));
+            }
+            // Allows the user to use MessagePack only.
+            _services.RemoveAll<IHubProtocol>();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:105]: shall we move this check to ServiceManagerBuilder.Build()

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/ServiceHubLifetimeManagerFactory.cs [hunk 2]
```diff
@@ -54,5 +56,14 @@ public ServiceHubLifetimeManagerFactory(IServiceProvider sp, IOptions<ServiceMan
                 default: throw new InvalidEnumArgumentException(nameof(ServiceManagerOptions.ServiceTransportType), (int)_options.ServiceTransportType, typeof(ServiceTransportType));
             }
         }
+
+        private void CheckHubProtocols()
+        {
+            var protocols = _serviceProvider.GetServices<IHubProtocol>().ToArray();
+            if (protocols.Length > 1 || protocols.Where(p => p.Name.Equals(Constants.Protocol.MessagePack)).Any())
+            {
+                throw new InvalidOperationException("ServiceManagerBuilder.WithHubProtocols method is not supported for transient(default) mode yet.");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/ServiceHubLifetimeManagerFactory.cs:65]:
```suggestion
                throw new InvalidOperationException("ServiceManagerBuilder.WithHubProtocol method is not supported for transient(default) mode yet.");
```

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs [hunk 3]
```diff
@@ -79,14 +80,40 @@ public ServiceManagerBuilder WithNewtonsoftJson(Action<NewtonsoftServiceHubProto
...
         }
 
+        /// <summary>
+        /// Sets the SignalR hub protocols to serialize messages sent to SignalR clients.
+        /// </summary>
+        /// <remarks><para>Currently it only works for <b>persistent</b> mode. The support for <b>transiet(default)</b> mode is to be done. </para> 
+        /// <para>Calling this method clears the default hub protocol.</para></remarks>
+        /// <param name="hubProtocols">Only the protocols named "json" or "messagepack" are allowed.</param>
+        /// <returns>The <see cref="ServiceHubContextBuilder"/> instance itself.</returns>
+        public ServiceManagerBuilder WithHubProtocol(IEnumerable<IHubProtocol> hubProtocols)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:98]: `WithHubProtocol(IHubProtocol hubProtocol)` and support multiple such calls?

Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:98]: The problem of `WithHubProtocol(IHubProtocol hubProtocol)` is that it only adds protocols, but cannot removes. If the user only wants MessagePack, they cannot achieve it.

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs (hunk 1 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:105]: Why do we want to move the check to `ServiceManagerBuilder.Build()`?

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs (hunk 3 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:98]: Then use `WithHubProtocols`

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs (hunk 1 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:105]: Build() is the time when everything is set. What if IHubProtocol is added after `WithHubProtocol`?

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs (hunk 3 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs]: how about `WithHubProtocols(params IHubProtocol[] ...)`

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs [hunk 4]
```diff
@@ -79,14 +80,40 @@ public ServiceManagerBuilder WithNewtonsoftJson(Action<NewtonsoftServiceHubProto
...
+        {
+            if (hubProtocols == null)
+            {
+                throw new ArgumentNullException(nameof(hubProtocols));
+            }
+            // Allows the user to use MessagePack only.
+            _services.RemoveAll<IHubProtocol>();
+            foreach (var hubProtocol in hubProtocols)
+            {
+                if (!hubProtocol.Name.Equals(Constants.Protocol.Json, StringComparison.OrdinalIgnoreCase) && !hubProtocol.Name.Equals(Constants.Protocol.MessagePack, StringComparison.OrdinalIgnoreCase))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:108]: is the protocol name case-insensitive?

File: src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/ServiceHubLifetimeManagerFactory.cs [hunk 5]
```diff
@@ -54,5 +56,14 @@ public ServiceHubLifetimeManagerFactory(IServiceProvider sp, IOptions<ServiceMan
                 default: throw new InvalidEnumArgumentException(nameof(ServiceManagerOptions.ServiceTransportType), (int)_options.ServiceTransportType, typeof(ServiceTransportType));
             }
         }
+
+        private void CheckHubProtocols()
+        {
+            var protocols = _serviceProvider.GetServices<IHubProtocol>().ToArray();
+            if (protocols.Length > 1 || protocols.Where(p => p.Name.Equals(Constants.Protocol.MessagePack)).Any())
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/ServiceHubLifetimeManagerFactory.cs:63]:
```suggestion
            if (protocols.Length > 1 || protocols.Any(p => p.Name.Equals(Constants.Protocol.MessagePack)))
```
case sensitive or insensitive here?

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs (hunk 3 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs]: updated

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs (hunk 4 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs:112]: Yes, consistent with SignalR implementation https://source.dot.net/#Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubProtocolResolver.cs,22

**Review by galvesribeiro — APPROVED** [review]
LGTM

File: src/Microsoft.Azure.SignalR.Management/Serialization/JsonObjectSerializerHubProtocol.cs [hunk 6]
```diff
@@ -57,15 +57,15 @@ internal sealed class JsonObjectSerializerHubProtocol : IHubProtocol
         private const string ProtocolName = "json";
         private const int ProtocolVersion = 1;
 
-        private readonly ObjectSerializer _objectSerializer;
```
Comment (Y-Sindo) [inline src/Microsoft.Azure.SignalR.Management/Serialization/JsonObjectSerializerHubProtocol.cs:60]: It has to be public later when implementing full support for MessagePack in transient mode. Make it public first for tests.

**Review by vicancy — APPROVED** [review]
Looks great to me!

**Comment by Y-Sindo** [comment]
@galvesribeiro The package is released.

**Comment by galvesribeiro** [comment]
@Y-Sindo thanks. But is it fixed for the Transient mode as well? Or it is just for persistent mode?

The case I've opened with the support is relative to Transient mode explicitly.

Thanks

**Comment by Y-Sindo** [comment]
> @Y-Sindo thanks. But is it fixed for the Transient mode as well? Or it is just for persistent mode?
> 
> The case I've opened with the support is relative to Transient mode explicitly.
> 
> Thanks

@galvesribeiro Not yet. The transient mode currently just sends JSON payload in REST API and lets the service translate JSON into MessagePack, which is why it doesn't work as you expected. To improve the behavior, we need to send MessagePack payload directly in the REST API, and the change will be bigger than the persistent mode and not that easy. To track this issue, please see #1745 .

**Comment by galvesribeiro** [comment]
Ok, then my ticket on the support is not fixed then...

**Comment by Y-Sindo** [comment]
@galvesribeiro Is it OK for you to switch to persistent mode as a workaround?

**Comment by Y-Sindo** [comment]
@galvesribeiro The [doc](https://github.com/Azure/azure-signalr/blob/dev/docs/management-sdk-guide.md) is also updated to remind users of the limitation of the MessagePack support for transient mode. We are also working actively to improve that.

**Comment by galvesribeiro** [comment]
Is it ok to have a single management connection as a singleton for the whole service assuming it has a very hight message volume to multiple target clients? In other words, is it thread-safe?

If that is ok, does it reconnect in case the connection is lost with SRS?

Is it queueing messages to be dispatched from multiple callers?

**Comment by Y-Sindo** [comment]
> Is it ok to have a single management connection as a singleton for the whole service assuming it has a very hight message volume to multiple target clients? In other words, is it thread-safe?

It is thread-safe. Each thread tries to wait for a lock to write message to the connection. 

> If that is ok, does it reconnect in case the connection is lost with SRS?

The connection will try to reconnect if it is lost. 

> Is it queueing messages to be dispatched from multiple callers?

No. As in the first answer, each caller will try to get a lock, and when the method returns, the message is already received by the service. During the interval when the connection is disconnected, you have to queue the message yourself.

**Pull request merged**